include/unifex/filter_stream.hpp

```diff
@@ -0,0 +1,388 @@
+  using sender = type;
+  StreamSender& stream_;
+  FilterFunc& filter_;
+
+  // value and error types just point to the ones from the stream sender
+  template <
+      template <typename...>
+      class Variant,
+      template <typename...>
+      class Tuple>
+  using value_types = typename next_sender_t<
+      StreamSender>::template value_types<Variant, Tuple>;
+
+  template <template <typename...> class Variant>
+  using error_types =
+      typename next_sender_t<StreamSender>::template error_types<Variant>;
+
+  static constexpr bool sends_done = false;
+
+  template <typename Receiver>
+  using operation_t = operation<StreamSender, FilterFunc, Receiver>;
+  template <typename Receiver>
+  using next_receiver_t = next_receiver<StreamSender, FilterFunc, Receiver>;
+  template <typename Receiver>
+  using error_cleanup_receiver_t =
+      error_cleanup_receiver<StreamSender, FilterFunc, Receiver>;
+  template <typename Receiver>
+  using done_cleanup_receiver_t =
+      done_cleanup_receiver<StreamSender, FilterFunc, Receiver>;
+
+  template(typename Self, typename Receiver)  //
+      (requires same_as<remove_cvref_t<Self>, type> AND receiver<Receiver> AND
+           sender_to<next_sender_t<StreamSender>, next_receiver_t<Receiver>> AND
+               sender_to<
+                   cleanup_sender_t<StreamSender>,
+                   error_cleanup_receiver_t<Receiver>> AND
+                   sender_to<
+                       cleanup_sender_t<StreamSender>,
+                       done_cleanup_receiver_t<Receiver>>)  //
+      friend operation_t<Receiver> tag_invoke(
+          tag_t<connect>, Self&& self, Receiver&& receiver) {
+    return operation_t<Receiver>{
+        self.stream_,
+        static_cast<Self&&>(self).filter_,
+        (Receiver &&) receiver};
+  }
+};
+}  // namespace _filter
+
+namespace _filter_stream {
+
+template <typename StreamSender, typename FilterFunc>
+struct _filter_stream {
+  struct type;
+};
+
+template <typename StreamSender, typename FilterFunc>
+struct _filter_stream<StreamSender, FilterFunc>::type {
+  StreamSender stream_;
+  FilterFunc filter_;
+
+  friend auto tag_invoke(tag_t<next>, type& s) {
+    return _filter::sender<StreamSender, FilterFunc>{s.stream_, s.filter_};
+  }
+
+  friend auto tag_invoke(tag_t<cleanup>, type& s) { return cleanup(s.stream_); }
+};
+
+struct _fn {
+  // TODO add a type requirement that filterFunc's return is boolean
+  template <typename StreamSender, typename FilterFunc>
+  auto operator()(StreamSender&& stream, FilterFunc&& filterFunc) const {
+    return typename _filter_stream<StreamSender, FilterFunc>::type{
+        (StreamSender &&) stream, (FilterFunc &&) filterFunc};
+  }
+
+  template <typename FilterFunc>
+  constexpr auto operator()(FilterFunc&& filterFunc) const
+      noexcept(is_nothrow_callable_v<tag_t<bind_back>, _fn, FilterFunc>)
+          -> bind_back_result_t<_fn, FilterFunc> {
+    return bind_back(*this, (FilterFunc &&) filterFunc);
```

Comment:
Just for my understanding, isn't `(FilterFunc &&) filterFunc` the same as `std::forward<FilterFunc>(filterFunc)`? The reference collapsing will end up producing the same reference type, based on my understanding. 

I have seen in the code a lot of usages of the `(T&&) t` style, when forwarding references are involved.

Reply:
It is the same, yes, and the existing uses of `(T&&) t` instead of `std::forward<T>(t)` are there because of concerns about compile times (the C-style cast doesn't require a function template instantiation but the `std::forward` version does). We decided recently, though (maybe 4 months ago?), that the difference in compile time didn't merit the loss in clarity and safety so we're planning to slowly update the code as we visit files, and to follow the new style in new code.